nio/crypto/sas.py

```diff
@@ -671,7 +671,8 @@ def receive_mac_event(self, event):
         if not self._event_ok(event):
             return
 
-        if self.state != SasState.key_received:
+        if ((self.state != SasState.key_received) or 
```

Comment:
So it seems that you somehow got into the `key_received` state either without calling `accept_verification` (this is a usage error) or without receiving an `m.key.verification.accept` event.

Adding a check that we are in the correct state (started or accepted) over [here](https://github.com/poljar/matrix-nio/blob/master/nio/crypto/sas.py#L637) sounds like a better fix since we're canceling the SAS flow before we can let the user see the emoji.

Reply:
I incorporated all your suggestions. 
I did some simple tests. 
More extensive testing would never hurt :)

Please look it over again.

See commit #2 in 47ef5c9bfae60672f09b0208fe77ec3265747c58